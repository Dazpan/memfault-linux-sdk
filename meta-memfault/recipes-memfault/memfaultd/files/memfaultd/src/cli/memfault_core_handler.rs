//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use super::cargs;

#[cfg(all(target_os = "linux", feature = "coredump"))]
pub fn main(args: cargs::CArgs) -> i32 {
    unsafe { memfaultc_sys::memfault_core_handler_main(args.argc(), args.argv()) }
}

#[cfg(not(all(target_os = "linux", feature = "coredump")))]
pub fn main(_args: cargs::CArgs) -> i32 {
    eprintln!("memfault-core-handler is not supported in this build");
    1
}
