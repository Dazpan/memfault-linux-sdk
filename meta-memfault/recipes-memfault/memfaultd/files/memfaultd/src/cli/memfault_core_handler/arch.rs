//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use cfg_if::cfg_if;

cfg_if! {
    if #[cfg(target_arch = "aarch64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.sp as usize
        }
    } else if #[cfg(target_arch = "x86_64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.rsp as usize
        }
    } else if #[cfg(target_arch = "arm")] {
        pub use libc::user_regs as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.arm_sp as usize
        }
    } else if #[cfg(target_arch = "x86")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.esp as usize
        }
    }
    else {
        // Provide dummy symbols for unsupported architectures. This is preferable to
        // a compile time error, as we want to be able to compile memfaultd for all
        // architectures, but we don't need register access for all of them. Currently
        // these registers are only used to filter out stack memory from coredumps.
        pub struct ElfGRegSet;
        pub fn get_stack_pointer(_regs: &ElfGRegSet) -> usize {
            0
        }
    }
}
