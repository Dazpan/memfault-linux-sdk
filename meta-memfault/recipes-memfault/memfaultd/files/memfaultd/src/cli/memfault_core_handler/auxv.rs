//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use scroll::{Pread, NATIVE};
use std::fmt::Debug;

#[cfg(target_pointer_width = "64")]
pub type AuxvUint = u64;

#[cfg(target_pointer_width = "32")]
pub type AuxvUint = u32;

#[derive(Eq, PartialEq)]
pub struct Auxv<'a> {
    data: &'a [u8],
}

#[derive(Eq, PartialEq, Debug)]
pub struct AuxvEntry {
    key: AuxvUint,
    value: AuxvUint,
}

impl<'a> Auxv<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self { data }
    }

    pub fn iter(&'a self) -> AuxvIterator<'a> {
        AuxvIterator::new(self)
    }
}

impl<'a> Debug for Auxv<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.iter().collect::<Vec<_>>().fmt(f)
    }
}

pub struct AuxvIterator<'a> {
    auxv: &'a Auxv<'a>,
    offset: usize,
}

impl<'a> AuxvIterator<'a> {
    fn new(auxv: &'a Auxv<'a>) -> Self {
        Self { auxv, offset: 0 }
    }
}

impl<'a> Iterator for AuxvIterator<'a> {
    type Item = AuxvEntry;

    fn next(&mut self) -> Option<Self::Item> {
        let mut read = || {
            self.auxv
                .data
                .gread_with::<AuxvUint>(&mut self.offset, NATIVE)
        };
        match (read(), read()) {
            (Ok(key), Ok(value)) => Some(AuxvEntry { key, value }),
            _ => None,
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use rstest::rstest;
    use scroll::IOwrite;
    use std::io::Cursor;

    #[rstest]
    // Empty auxv:
    #[case(
        vec![],
        vec![]
    )]
    // Happy path:
    #[case(
        vec![1, 2, 3, 4],
        vec![AuxvEntry { key: 1, value: 2 }, AuxvEntry { key: 3, value: 4 }]
    )]
    // Partial entry at the end is ignored:
    #[case(
        vec![1, 2, 3],
        vec![AuxvEntry { key: 1, value: 2 }]
    )]
    fn test_auxv_iterator(#[case] values: Vec<AuxvUint>, #[case] expected: Vec<AuxvEntry>) {
        let mut cursor = Cursor::new(vec![]);
        for value in values {
            cursor.iowrite_with::<AuxvUint>(value, NATIVE).unwrap();
        }

        let buffer = cursor.into_inner();
        let auxv = Auxv::new(buffer.as_slice());
        assert_eq!(auxv.iter().collect::<Vec<_>>(), expected);
    }
}
