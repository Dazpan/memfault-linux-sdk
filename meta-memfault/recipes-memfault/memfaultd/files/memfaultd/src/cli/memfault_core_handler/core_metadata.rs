//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::time::SystemTime;

use crate::build_info::VERSION;

use ciborium::{cbor, into_writer};
use eyre::Result;

use super::core_elf_note::build_elf_note;

const NOTE_NAME: &str = "Memfault\0";
const NOTE_TYPE: u32 = 0x4154454d;
const MEMFAULT_CORE_ELF_METADATA_SCHEMA_VERSION_V1: u32 = 1;

enum MemfaultCoreElfMetadataKey {
    SchemaVersion = 1,
    LinuxSdkVersion = 2,
    CapturedTime = 3,
    DeviceSerial = 4,
    HardwareVersion = 5,
    SoftwareType = 6,
    SoftwareVersion = 7,
}

#[derive(Debug)]
pub struct CoredumpMetadata {
    pub device_id: String,
    pub hardware_version: String,
    pub software_type: String,
    pub software_version: String,
    pub sdk_version: String,
    pub captured_time_epoch_s: u64,
}

impl From<&crate::config::Config> for CoredumpMetadata {
    fn from(config: &crate::config::Config) -> Self {
        Self {
            device_id: config.device_info.device_id.clone(),
            hardware_version: config.device_info.hardware_version.clone(),
            software_type: config.config_file.software_type.clone(),
            software_version: config.config_file.software_version.clone(),
            sdk_version: VERSION.to_string(),
            captured_time_epoch_s: SystemTime::now()
                .duration_since(SystemTime::UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs(),
        }
    }
}

pub fn serialize_metadata_as_map(metadata: &CoredumpMetadata) -> Result<Vec<u8>> {
    let cbor_val = cbor!({
        MemfaultCoreElfMetadataKey::SchemaVersion as u32 => MEMFAULT_CORE_ELF_METADATA_SCHEMA_VERSION_V1,
        MemfaultCoreElfMetadataKey::LinuxSdkVersion as u32 => metadata.sdk_version,
        MemfaultCoreElfMetadataKey::CapturedTime as u32 => metadata.captured_time_epoch_s,
        MemfaultCoreElfMetadataKey::DeviceSerial as u32 => metadata.device_id,
        MemfaultCoreElfMetadataKey::HardwareVersion as u32 => metadata.hardware_version,
        MemfaultCoreElfMetadataKey::SoftwareType as u32 => metadata.software_type,
        MemfaultCoreElfMetadataKey::SoftwareVersion as u32 => metadata.software_version,
    })?;

    let mut buffer = Vec::new();
    into_writer(&cbor_val, &mut buffer)?;

    Ok(buffer)
}

pub fn write_memfault_note(metadata: &CoredumpMetadata) -> Result<Vec<u8>> {
    let description_buffer = serialize_metadata_as_map(metadata)?;

    build_elf_note(NOTE_NAME, &description_buffer, NOTE_TYPE)
}

#[cfg(test)]
mod test {
    use ciborium::{from_reader, Value};

    use super::*;

    #[test]
    fn test_serialize_metadata_as_map() {
        let metadata = CoredumpMetadata {
            device_id: "12345678".to_string(),
            hardware_version: "evt".to_string(),
            software_type: "main".to_string(),
            software_version: "1.0.0".to_string(),
            sdk_version: "SDK_VERSION".to_string(),
            captured_time_epoch_s: 1234,
        };

        let map = serialize_metadata_as_map(&metadata).unwrap();
        let deser_map: Value = from_reader(map.as_slice()).unwrap();
        insta::assert_debug_snapshot!(deser_map);
        assert_eq!(map.len(), 48);
    }
}
