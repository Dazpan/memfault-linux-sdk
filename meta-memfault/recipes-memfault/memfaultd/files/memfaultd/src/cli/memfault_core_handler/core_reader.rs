//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::cmp::Ordering;
use std::io::{copy, sink, Read};

use crate::cli::memfault_core_handler::elf;

use elf::header::{Header, EI_CLASS, ELFCLASS, ELFMAG, EV_CURRENT, SELFMAG, SIZEOF_EHDR};
use elf::program_header::{ProgramHeader, SIZEOF_PHDR};
use eyre::{eyre, Result};
use log::trace;

pub trait CoreReader {
    /// Reads program headers from the input stream
    fn read_program_headers(&mut self) -> Result<Vec<ProgramHeader>>;

    /// Reads segment data from the input stream
    fn read_segment_data(&mut self, program_header: &ProgramHeader) -> Result<Vec<u8>>;
}

// Reads ELF headers and segments from a core file
pub struct CoreReaderImpl<R: Read> {
    input_stream: R,
    elf_header: Header,
    // Note: we track the cursor position manually here instead of requiring `Seek` so we can
    // support reading from stdin which doesn't implement `Seek`.
    cursor: usize,
}

impl<R: Read> CoreReader for CoreReaderImpl<R> {
    fn read_program_headers(&mut self) -> Result<Vec<ProgramHeader>> {
        let header_size = self.elf_header.e_phnum as usize * SIZEOF_PHDR;
        let mut header_buf = vec![0; header_size];
        self.skip_to_offset(self.elf_header.e_phoff as usize)?;

        let bytes_read = self.read_input(&mut header_buf)?;
        if bytes_read != header_size {
            return Err(eyre!("Failed to read program header table"));
        }

        let mut program_headers =
            ProgramHeader::from_bytes(&header_buf, self.elf_header.e_phnum as usize);

        // Sort, just in case the program headers are not sorted by offset.
        // Otherwise the read_segment_data() calls later may fail.
        program_headers.sort_by_key(|ph| ph.p_offset);

        Ok(program_headers)
    }

    fn read_segment_data(&mut self, program_header: &ProgramHeader) -> Result<Vec<u8>> {
        self.skip_to_offset(program_header.p_offset as usize)?;

        let mut buf = vec![0; program_header.p_filesz as usize];
        self.read_input(&mut buf)?;

        Ok(buf)
    }
}

impl<R: Read> CoreReaderImpl<R> {
    /// Creates an instance of `CoreReader` from an input stream
    pub fn new(mut input_stream: R) -> Result<Self> {
        let mut header_buf = [0u8; SIZEOF_EHDR];
        input_stream.read_exact(&mut header_buf)?;

        let elf_header = *Header::from_bytes(&header_buf);
        if !Self::verify_elf_header(&elf_header) {
            return Err(eyre!("Invalid ELF header"));
        }

        Ok(CoreReaderImpl {
            input_stream,
            elf_header,
            cursor: SIZEOF_EHDR,
        })
    }

    fn verify_elf_header(header: &Header) -> bool {
        &header.e_ident[0..SELFMAG] == ELFMAG
            && header.e_ident[EI_CLASS] == ELFCLASS
            && header.e_version == EV_CURRENT as u32
            && header.e_ehsize == SIZEOF_EHDR as u16
            && header.e_phentsize == SIZEOF_PHDR as u16
    }

    pub fn elf_header(&self) -> Header {
        self.elf_header
    }

    // Skip bytes in the stream if cursor is less than requested offset.
    fn skip_to_offset(&mut self, offset: usize) -> Result<()> {
        match self.cursor.cmp(&offset) {
            Ordering::Less => {
                // Skip to the segment
                let skip_size = offset - self.cursor;
                trace!("Skipping {} bytes from cursor {}", skip_size, self.cursor);
                self.skip_bytes(skip_size)?;
            }
            Ordering::Equal => {
                trace!("Already at requested offset");
            }
            Ordering::Greater => {
                return Err(eyre!("Cursor already past requested offset"));
            }
        }

        Ok(())
    }

    /// Read from input stream and update cursor position
    fn read_input(&mut self, buf: &mut [u8]) -> Result<usize> {
        self.input_stream.read_exact(buf)?;
        self.cursor += buf.len();

        Ok(buf.len())
    }

    /// Consume specified number of bytes from the input stream
    fn skip_bytes(&mut self, bytes: usize) -> Result<()> {
        self.cursor += copy(
            &mut self.input_stream.by_ref().take(bytes as u64),
            &mut sink(),
        )? as usize;

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::cli::memfault_core_handler::test_utils::build_test_header;

    use rstest::rstest;
    use scroll::Pwrite;

    #[rstest]
    #[case(0)]
    #[case(1024)] // Test with padding between header and program headers
    fn test_read_program_headers(#[case] ph_offset: usize) {
        let mut elf_header = build_test_header();
        elf_header.e_phnum = 2;
        elf_header.e_phoff = (SIZEOF_EHDR + ph_offset).try_into().unwrap();
        let load_program_header = ProgramHeader {
            p_type: elf::program_header::PT_LOAD,
            p_vaddr: 0,
            ..Default::default()
        };
        let note_program_header = ProgramHeader {
            p_type: elf::program_header::PT_NOTE,
            p_offset: 0x1000,
            ..Default::default()
        };

        // Build ELF input stream
        let mut input_stream = vec![0; SIZEOF_EHDR + 2 * SIZEOF_PHDR + ph_offset];
        input_stream
            .pwrite_with(elf_header, 0, scroll::NATIVE)
            .unwrap();
        input_stream
            .pwrite_with(load_program_header, SIZEOF_EHDR + ph_offset, scroll::NATIVE)
            .unwrap();
        input_stream
            .pwrite_with(
                note_program_header,
                SIZEOF_EHDR + SIZEOF_PHDR + ph_offset,
                scroll::NATIVE,
            )
            .unwrap();

        // Verify headers are read correctly
        let mut reader = CoreReaderImpl::new(input_stream.as_slice()).unwrap();
        let program_headers = reader.read_program_headers().unwrap();
        assert_eq!(program_headers.len(), 2);
        assert_eq!(program_headers[0], load_program_header);
        assert_eq!(program_headers[1], note_program_header);
    }

    #[rstest]
    #[case(0, 1024)]
    #[case(1024, 1024)]
    fn test_read_segment_data(#[case] offset: usize, #[case] size: usize) {
        const TEST_BYTE: u8 = 0x42;

        let elf_header = build_test_header();
        let offset = offset + SIZEOF_EHDR;
        let note_program_header = ProgramHeader {
            p_type: elf::program_header::PT_NOTE,
            p_offset: offset.try_into().unwrap(),
            p_filesz: size.try_into().unwrap(),
            ..Default::default()
        };

        let mut input_stream = vec![0u8; offset + size];
        input_stream
            .pwrite_with(elf_header, 0, scroll::NATIVE)
            .unwrap();
        input_stream[offset..(offset + size)].fill(TEST_BYTE);

        let mut reader = CoreReaderImpl::new(input_stream.as_slice()).unwrap();
        let segment_data = reader.read_segment_data(&note_program_header).unwrap();

        assert_eq!(segment_data, input_stream[offset..(offset + size)]);
    }
}
