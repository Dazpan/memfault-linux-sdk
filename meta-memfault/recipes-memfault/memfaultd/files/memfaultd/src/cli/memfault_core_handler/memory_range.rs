//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use crate::cli::memfault_core_handler::ElfPtrSize;
use std::cmp::max;

/// Convenience struct to manage memory address ranges
#[derive(Debug, PartialEq, Eq)]
pub struct MemoryRange {
    pub start: ElfPtrSize,
    pub end: ElfPtrSize,
}

impl MemoryRange {
    pub fn new(start: ElfPtrSize, end: ElfPtrSize) -> Self {
        Self { start, end }
    }

    pub fn from_start_and_size(start: ElfPtrSize, size: ElfPtrSize) -> Self {
        Self {
            start,
            end: start + size,
        }
    }

    /// Returns true if the two ranges overlap or touch (end-inclusive checking!).
    pub fn overlaps(&self, other: &Self) -> bool {
        self.start <= other.end && self.end >= other.start
    }

    pub fn size(&self) -> ElfPtrSize {
        self.end - self.start
    }
}

/// Merge overlapping memory ranges.
///
/// This is used to merge memory ranges before turning them into PT_LOAD program
/// headers.
pub fn merge_memory_ranges(mut ranges: Vec<MemoryRange>) -> Vec<MemoryRange> {
    // First, sort by start address. This lets us merge overlapping ranges in a single pass
    // by knowing that we only need to check the last range in the merged list.
    ranges.sort_by_key(|r| r.start);

    // Next, iterate over the sorted ranges and merge overlapping ranges. If the current range
    // overlaps with the last range in the merged list, we extend the last range to include the
    // current range. Otherwise, we add the current range to the merged list.
    let mut merged_ranges: Vec<MemoryRange> = Vec::new();
    for range in ranges {
        if let Some(last) = merged_ranges.last_mut() {
            if last.overlaps(&range) {
                last.end = max(last.end, range.end);
                continue;
            }
        }
        merged_ranges.push(range);
    }

    merged_ranges
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use super::*;

    #[rstest]
    // Two ranges with matching boundaries
    #[case(
        vec![MemoryRange::new(0x1000, 0x2000), MemoryRange::new(0x2000, 0x3000)],
        vec![MemoryRange::new(0x1000, 0x3000)],
    )]
    // Two ranges with overlapping boundaries
    #[case(
        vec![MemoryRange::new(0x1000, 0x2000), MemoryRange::new(0x1500, 0x3000)],
        vec![MemoryRange::new(0x1000, 0x3000)],
    )]
    // Two ranges with non-overlapping boundaries
    #[case(
        vec![MemoryRange::new(0x1000, 0x2000), MemoryRange::new(0x3000, 0x4000)],
        vec![MemoryRange::new(0x1000, 0x2000), MemoryRange::new(0x3000, 0x4000)],
    )]
    // Three overlapping regions, unsorted
    #[case(
        vec![
            MemoryRange::new(0x1500, 0x3000),
            MemoryRange::new(0x1000, 0x2000),
            MemoryRange::new(0x3000, 0x5000),
        ],
        vec![MemoryRange::new(0x1000, 0x5000)]
    )]
    fn test_memory_range_merge(
        #[case] input: Vec<MemoryRange>,
        #[case] expected: Vec<MemoryRange>,
    ) {
        let merged = merge_memory_ranges(input);
        assert_eq!(merged, expected);
    }
}
