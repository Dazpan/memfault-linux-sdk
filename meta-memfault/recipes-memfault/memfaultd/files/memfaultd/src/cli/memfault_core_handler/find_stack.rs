//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::cmp::min;

use crate::cli::memfault_core_handler::arch::{get_stack_pointer, ElfGRegSet};
use crate::cli::memfault_core_handler::elf;
use crate::cli::memfault_core_handler::memory_range::MemoryRange;
use crate::cli::memfault_core_handler::ElfPtrSize;

use elf::program_header::ProgramHeader;

/// Attempts to find a MemoryRange for the stack based on the supplied register set.
/// The returned range is bound by the max_thread_size and the end of the segment in which
/// the stack is found. The assumption herein is that an anonymous memory mapping is created as
/// stack (and used exclusively as such). If the stack pointer is not contained in any segment,
/// None is returned.
pub fn find_stack(
    regs: &ElfGRegSet,
    program_headers: &[ProgramHeader],
    max_thread_size: usize,
) -> Option<MemoryRange> {
    let stack_pointer = get_stack_pointer(regs) as ElfPtrSize;

    // Iterate over all PT_LOAD segments and find the one that contains the stack
    // pointer. If the stack pointer is not contained in any PT_LOAD segment, we
    // ignore the thread.
    //
    // NOTE: This is an M*N operation, but both M(#segments) and N(#threads) are
    // likely quite small so this should be fine.
    for header in program_headers {
        let region = header.p_vaddr..header.p_vaddr.saturating_add(header.p_memsz);

        if region.contains(&stack_pointer) {
            // MFLT-11631 Handle upward stack growth in Threads memory selector
            let stack_size = min(
                region.end.saturating_sub(stack_pointer),
                max_thread_size as ElfPtrSize,
            );
            let stack_base = stack_pointer + stack_size;

            return Some(MemoryRange::new(stack_pointer, stack_base));
        }
    }

    None
}
