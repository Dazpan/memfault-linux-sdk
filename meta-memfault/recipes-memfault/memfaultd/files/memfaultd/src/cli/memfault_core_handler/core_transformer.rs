//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::io::{Read, Seek, SeekFrom};
use std::mem::size_of;

use crate::cli::memfault_core_handler::core_elf_note::{iterate_elf_notes, ElfNote, MappedFile};
use crate::cli::memfault_core_handler::core_metadata::{write_memfault_note, CoredumpMetadata};
use crate::cli::memfault_core_handler::core_reader::CoreReader;
use crate::cli::memfault_core_handler::core_writer::{CoreWriter, SegmentData};
use crate::cli::memfault_core_handler::elf;
use crate::cli::memfault_core_handler::memory_range::{merge_memory_ranges, MemoryRange};
use crate::cli::memfault_core_handler::ElfPtrSize;
use crate::config::CoredumpCaptureStrategy;

use crate::cli::memfault_core_handler::find_elf_headers::find_elf_headers_and_build_id_note_ranges;
use crate::cli::memfault_core_handler::find_stack::find_stack;
use elf::program_header::{ProgramHeader, PT_LOAD, PT_NOTE};
use eyre::{eyre, Result};
use log::warn;

#[derive(Debug)]
pub struct CoreTransformerOptions {
    pub max_size: usize,
    pub capture_strategy: CoredumpCaptureStrategy,
}

/// Reads segments from core elf stream and memory stream and builds a core new elf file.
pub struct CoreTransformer<R, W, P>
where
    R: CoreReader,
    W: CoreWriter,
    P: Read + Seek,
{
    core_reader: R,
    core_writer: W,
    proc_mem_stream: P,
    metadata: CoredumpMetadata,
    options: CoreTransformerOptions,
}

impl<R, W, P> CoreTransformer<R, W, P>
where
    R: CoreReader,
    W: CoreWriter,
    P: Read + Seek,
{
    /// Creates an instance of `CoreTransformer` from an input stream and output stream
    pub fn new(
        core_reader: R,
        core_writer: W,
        proc_mem_stream: P,
        options: CoreTransformerOptions,
        metadata: CoredumpMetadata,
    ) -> Result<Self> {
        Ok(Self {
            core_reader,
            core_writer,
            proc_mem_stream,
            metadata,
            options,
        })
    }

    /// Reads segments from core elf stream and memory stream and builds a new elf file
    ///
    /// Reads all PT_LOAD and PT_NOTE program headers and their associated segments from the core.
    /// The memory for each PT_LOAD segment will be fetched from `/proc/<pid>/mem` and the
    /// resulting elf file will be written to `output_stream`.
    pub fn run_transformer(&mut self) -> Result<()> {
        let program_headers = self.core_reader.read_program_headers()?;
        let all_notes = self.read_all_note_segments(&program_headers);

        let segments_to_capture = match self.options.capture_strategy {
            CoredumpCaptureStrategy::KernelSelection => {
                self.kernel_selection_segments(&program_headers)
            }
            CoredumpCaptureStrategy::Threads { max_thread_size } => {
                self.threads_segments(&program_headers, &all_notes, max_thread_size)
            }
        };

        // Always copy over all note segments, regardless of the capturing strategy:
        for (ph, data) in all_notes {
            self.core_writer.add_segment(*ph, SegmentData::Buffer(data));
        }

        for ph in segments_to_capture {
            self.core_writer.add_segment(ph, SegmentData::ProcessMemory);
        }

        self.add_memfault_note()?;
        self.check_output_size()?;
        self.core_writer.write()?;

        Ok(())
    }

    /// Add note segments from the core elf to the output elf, verbatim.
    fn read_all_note_segments<'a>(
        &mut self,
        program_headers: &'a [ProgramHeader],
    ) -> Vec<(&'a ProgramHeader, Vec<u8>)> {
        program_headers
            .iter()
            .filter_map(|ph| match ph.p_type {
                PT_NOTE => match self.core_reader.read_segment_data(ph) {
                    Ok(data) => Some((ph, data)),
                    _ => None,
                },
                _ => None,
            })
            .collect::<Vec<_>>()
    }

    /// All load segments from the original/input core elf as provided by the kernel, verbatim.
    fn kernel_selection_segments(
        &mut self,
        program_headers: &[ProgramHeader],
    ) -> Vec<ProgramHeader> {
        program_headers
            .iter()
            .filter_map(|ph| match ph.p_type {
                PT_LOAD => Some(*ph),
                _ => None,
            })
            .collect::<Vec<_>>()
    }

    /// Synthesizes segments for the stacks of all threads and adds them to the output, as well as
    /// any mandatory segments (build id notes, r_debug, etc.).
    fn threads_segments(
        &mut self,
        program_headers: &[ProgramHeader],
        all_notes: &[(&ProgramHeader, Vec<u8>)],
        max_thread_size: usize,
    ) -> Vec<ProgramHeader> {
        let parsed_notes = all_notes
            .iter()
            .flat_map(|(_, data)| iterate_elf_notes(data))
            .collect::<Vec<_>>();

        let mut mem_ranges = Vec::new();
        for note in &parsed_notes {
            match note {
                ElfNote::ProcessStatus(s) => {
                    if let Some(stack) = find_stack(&s.pr_reg, program_headers, max_thread_size) {
                        mem_ranges.push(stack);
                    } else {
                        warn!("Failed to collect stack for thread: {}", s.pr_pid);
                    }
                }
                ElfNote::File(f) => {
                    mem_ranges.extend(f.iter().filter(|f| f.page_offset == 0).flat_map(|file| {
                        match self.elf_metadata_ranges_for_mapped_file(file) {
                            Ok(ranges) => ranges,
                            Err(e) => {
                                warn!(
                                    "Failed to collect metadata for {} @ {:#x}: {}",
                                    file.path
                                        .map(|p| p.display().to_string())
                                        .unwrap_or_else(|| "???".into()),
                                    file.start_addr,
                                    e
                                );
                                vec![]
                            }
                        }
                    }));
                }
                _ => { /* TODO */ }
            }
        }

        // Merge overlapping memory ranges and turn them into PT_LOAD program headers. As a
        // side-effect, this will also sort the program headers by vaddr.
        let merged_ranges = merge_memory_ranges(mem_ranges);
        merged_ranges.into_iter().map(ProgramHeader::from).collect()
    }

    fn elf_metadata_ranges_for_mapped_file(
        &mut self,
        file: &MappedFile,
    ) -> Result<Vec<MemoryRange>> {
        let vaddr_base = file.start_addr as ElfPtrSize;
        self.proc_mem_stream
            .seek(SeekFrom::Start(vaddr_base as u64))?;
        find_elf_headers_and_build_id_note_ranges(vaddr_base, &mut self.proc_mem_stream)
    }

    /// Check if the output file size exceeds the max size available
    fn check_output_size(&self) -> Result<()> {
        let output_size = self.core_writer.calc_output_size();
        if output_size > self.options.max_size {
            return Err(eyre!(
                "Core file size {} exceeds max size {}",
                output_size,
                self.options.max_size
            ));
        }

        Ok(())
    }

    /// Add memfault note to the core elf
    ///
    /// Contains CBOR encoded information about the system capturing the coredump. See
    /// [`CoredumpMetadata`] for more information.
    fn add_memfault_note(&mut self) -> Result<()> {
        let note_data = write_memfault_note(&self.metadata)?;
        let program_header = ProgramHeader {
            p_type: PT_NOTE,
            p_filesz: note_data.len().try_into()?,
            ..Default::default()
        };
        self.core_writer
            .add_segment(program_header, SegmentData::Buffer(note_data));

        Ok(())
    }
}

impl From<MemoryRange> for ProgramHeader {
    fn from(range: MemoryRange) -> Self {
        ProgramHeader {
            p_type: PT_LOAD,
            p_vaddr: range.start,
            p_filesz: range.size(),
            p_memsz: range.size(),
            p_align: size_of::<ElfPtrSize>() as ElfPtrSize,
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod test {
    use crate::cli::memfault_core_handler::test_utils::{FakeProcMem, MockCoreWriter};
    use crate::test_utils::setup_logger;
    use crate::{
        cli::memfault_core_handler::core_reader::CoreReaderImpl, test_utils::set_snapshot_suffix,
    };
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use std::fs::File;
    use std::path::PathBuf;

    use super::*;

    #[rstest]
    #[case("kernel_selection", CoredumpCaptureStrategy::KernelSelection)]
    #[case("threads_32k", CoredumpCaptureStrategy::Threads { max_thread_size: 32 * 1024 })]
    fn test_transform(
        #[case] test_case_name: &str,
        #[case] capture_strategy: CoredumpCaptureStrategy,
        _setup_logger: (),
    ) {
        let input_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/elf-core-runtime-ld-paths.elf");
        let input_stream = File::open(&input_path).unwrap();
        let proc_mem_stream = FakeProcMem::new_from_path(&input_path).unwrap();
        let opts = CoreTransformerOptions {
            max_size: 1024 * 1024,
            capture_strategy,
        };
        let metadata = CoredumpMetadata {
            device_id: "12345678".to_string(),
            hardware_version: "evt".to_string(),
            software_type: "main".to_string(),
            software_version: "1.0.0".to_string(),
            sdk_version: "SDK_VERSION".to_string(),
            captured_time_epoch_s: 1234,
        };

        let core_reader = CoreReaderImpl::new(input_stream).unwrap();
        let mut segments = vec![];
        let mock_core_writer = MockCoreWriter::new(&mut segments);
        let mut transformer = CoreTransformer::new(
            core_reader,
            mock_core_writer,
            proc_mem_stream,
            opts,
            metadata,
        )
        .unwrap();

        transformer.run_transformer().unwrap();

        // Omit the actual data from the notes:
        let segments = segments
            .iter()
            .map(|(ph, seg)| {
                let seg = match seg {
                    SegmentData::ProcessMemory => SegmentData::ProcessMemory,
                    SegmentData::Buffer(_) => SegmentData::Buffer(vec![]),
                };
                (ph, seg)
            })
            .collect::<Vec<_>>();

        set_snapshot_suffix!("{}", test_case_name);
        assert_debug_snapshot!(segments);
    }
}
