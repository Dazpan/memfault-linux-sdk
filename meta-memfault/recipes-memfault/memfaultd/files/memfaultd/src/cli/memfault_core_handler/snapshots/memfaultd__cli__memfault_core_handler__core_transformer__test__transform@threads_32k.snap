---
source: memfaultd/src/cli/memfault_core_handler/core_transformer.rs
expression: segments
---
[
    (
        ProgramHeader {
            p_type: "PT_NOTE",
            p_flags: 0x0,
            p_offset: 0x660,
            p_vaddr: 0x0,
            p_paddr: 0x0,
            p_filesz: 0xdec,
            p_memsz: 0x0,
            p_align: 0,
        },
        Buffer(
            [],
        ),
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x5587ae8bd000,
            p_paddr: 0x0,
            p_filesz: 0x39c,
            p_memsz: 0x39c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38cb4000,
            p_paddr: 0x0,
            p_filesz: 0x3b4,
            p_memsz: 0x3b4,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38e99000,
            p_paddr: 0x0,
            p_filesz: 0x30c,
            p_memsz: 0x30c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7f6e38f7a000,
            p_paddr: 0x0,
            p_filesz: 0x25c,
            p_memsz: 0x25c,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_LOAD",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x7ffe4ba7a3c0,
            p_paddr: 0x0,
            p_filesz: 0x1c40,
            p_memsz: 0x1c40,
            p_align: 8,
        },
        ProcessMemory,
    ),
    (
        ProgramHeader {
            p_type: "PT_NOTE",
            p_flags: 0x0,
            p_offset: 0x0,
            p_vaddr: 0x0,
            p_paddr: 0x0,
            p_filesz: 0x48,
            p_memsz: 0x0,
            p_align: 0,
        },
        Buffer(
            [],
        ),
    ),
]
